src/osd/PrimaryLogPG.cc

```diff
@@ -1638,19 +1638,29 @@ void PrimaryLogPG::calc_trim_to()
 	cct->_conf->osd_pg_log_trim_max >= cct->_conf->osd_pg_log_trim_min) {
       return;
     }
-    list<pg_log_entry_t>::const_iterator it = pg_log.get_log().log.begin();
-    eversion_t new_trim_to;
-    for (uint64_t i = 0; i < num_to_trim; ++i) {
-      new_trim_to = it->version;
-      ++it;
-      if (new_trim_to >= limit) {
-	new_trim_to = limit;
-        dout(10) << "calc_trim_to trimming to limit: " << limit << dendl;
-	break;
+    auto it = pg_log.get_log().log.begin(); // oldest log entry
+    auto rit = pg_log.get_log().log.rbegin();
+    eversion_t by_n_to_keep; // start from tail
+    eversion_t by_n_to_trim = eversion_t::max(); // start from head
+    for (size_t i = 0; it != pg_log.get_log().log.end(); ++it, ++rit) {
+      i++;
+      if (i > target && by_n_to_keep == eversion_t()) {
+        by_n_to_keep = rit->version;
+      }
+      if (i >= num_to_trim && by_n_to_trim == eversion_t::max()) {
+        by_n_to_trim = it->version;
+        if (by_n_to_keep != eversion_t()) {
+          break;
+        }
       }
     }
-    dout(10) << "calc_trim_to " << pg_trim_to << " -> " << new_trim_to << dendl;
-    pg_trim_to = new_trim_to;
+
+    if (by_n_to_keep == eversion_t()) {
+      return;
+    }
```

Comment:
`num_to_trim` here is the upper bound of the number of entries to trim. now, the only concern is that the actual num to trim could be less than `osd_pg_log_trim_min`. in that case, `by_n_to_trim` is `eversion_t::max()`.

not sure how often it happens though, or if it really happens in real world, and if we care about this, and want to avoid it. we need to note down `i` when `by_n_to_keep` is set, and the total number of `pg_log.get_log().log` to calculate the number to trim in case `by_n_to_trim` is not set.

Reply:
I don't understand. Wouldn't pg_trim_to also be hard-caped at by_n_to_keep or head?

Reply:
it has nothing to do with `by_n_to_keep` , but is related the number to trim, there is chance that we never reach the `num_to_trim` when iterating thru the log. in that case, the number to trim could be less than `osd_pg_log_trim_min`.

Reply:
I see... I guess we can leave it as it is? Because there is no O(1) way to get the log size and that is why this (complicated)  fix is needed...